### PR TITLE
[ORCT-144] Update periods API

### PR DESCRIPTION
**Pull request by Ehevi**
## I have a [JIRA ticket](https://alice.its.cern.ch/jira/projects/ORCT/issues/ORCT-144?filter=allopenissues)
- [x] branch and/or PR name(s) include(s) JIRA ID
- [ ] issue has "Fix version" assigned
- [ ] issue "Status" is set to "In review"
- [x] PR labels are selected

### Notable changes for users:
- add year field to the periods view
- add a list of distinct run energies to the period view

### Notable changes for developers:
- use the new API endpoint for the periods view
- rename `pager` to `obsoletePager` -> from now please use the new pager only
- move `userPreferences` up in the model hierarchy
- remove unused `delay` function

### CODE REVIEW

File: app/public/Model.js [hunk 1]
```diff
@@ -49,6 +57,22 @@ export default class Model extends Observable {
...
         this.login('physicist');
+
+        this.handleLocationChange();
+    }
+
+    /**
+     * Delegates sub-model actions depending on new location of the page
+     * @returns {void}
+     */
+    async handleLocationChange() {
```
Comment (Ehevi) [inline app/public/Model.js:68]: This function is meant to handle all the views once they are implemented in a similar way to periods.

File: app/public/components/messagePanel/messages.js [hunk 2]
```diff
@@ -19,26 +19,47 @@ import spinner from '../common/spinner.js';
 const goBack = 'Go back';
 const nothingFound = 'Nothing found';
 
-const requestButton = (model) => h('button.btn.btn-primary.m3', {
+const obsoleteRequestButton = (model) => h('button.btn.btn-primary.m3', {
```
Comment (Ehevi) [inline app/public/components/messagePanel/messages.js:22]: Meant to be removed once all the requests are done with `fetchCurrentPageData()`

File: app/public/components/table/pagination/PaginationModel.js
Comment (Ehevi) [inline app/public/components/table/pagination/PaginationModel.js:1]: Strongly inspired by the BK [PaginationModel](https://github.com/AliceO2Group/Bookkeeping/blob/main/lib/public/components/Pagination/PaginationModel.js), but without infinity scroll support.

File: app/public/model/UserPreferences.js [hunk 3]
```diff
@@ -39,8 +39,9 @@ export default class UserPreferences extends Observable {
         this.detectorList = RCT.detectors.reduce((acc, detector) => ({ ...acc, [detector]: true }), {});
     }
 
-    setRowsOnSite(rowsOnSite) {
-        this.rowsOnSite = rowsOnSite;
+    setItemsPerPage(itemsPerPage) {
+        this.rowsOnSite = itemsPerPage;
```
Comment (Ehevi) [inline app/public/model/UserPreferences.js:43]: Some time ago we were told to change `page` to `site` (prev version: `rowsPerPage`, now `rowsOnSite`), but in the BK I've seen `page` being used in this context everywhere. Should I change all the related variables this way?

Comment (Ehevi) [reply app/public/model/UserPreferences.js:43]: The lines below need to stay for retro-compatibility reasons (other views still need them for changing `rowsOnSite`)

File: app/public/model/data/FetchedDataManager.js [hunk 4]
```diff
@@ -51,7 +51,7 @@ export default class FetchedDataManager {
         if (url === null) {
             url = this.router.getUrl();
             if (!url.searchParams.has(dataReqParams.rowsOnSite)) {
-                url = new URL(`${url.href}&${dataReqParams.rowsOnSite}=${this.model.userPreferences.rowsOnSite}`);
+                url = new URL(`${url.href}&${dataReqParams.rowsOnSite}=${this.model.parent.userPreferences.rowsOnSite}`);
```
Comment (Ehevi) [inline app/public/model/data/FetchedDataManager.js:54]: `userPreferences` have been moved up in the model hierarchy, hence the change here. The whole `FetchedDataManager` file is meant to be removed in the nearest future (once all the views use the new endpoints).

File: app/public/utils/fetch/getRemoteData.js
Comment (Ehevi) [inline app/public/utils/fetch/getRemoteData.js:1]: copy-paste of the [BK file](https://github.com/AliceO2Group/Bookkeeping/blob/main/lib/public/utilities/fetch/getRemoteData.js)

File: app/public/utils/fetch/jsonFetch.js
Comment (Ehevi) [inline app/public/utils/fetch/jsonFetch.js:1]: copy-paste of the [BK file](https://github.com/AliceO2Group/Bookkeeping/blob/main/lib/public/utilities/fetch/jsonFetch.js)

File: app/public/views/periods/overview/periodsContent.js [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+
+const applicableDataActions = {
+    [dataActions.hide]: true,
+    [dataActions.reload]: true,
+    [dataActions.downloadCSV]: true,
+    [dataActions.copyLink]: true,
+    [dataActions.showFilteringPanel]: true,
+};
+
+export default function periodsContent(periodsModel, model) {
```
Comment (Ehevi) [inline app/public/views/periods/overview/periodsContent.js:42]: `model` is meant to be removed from the list of arguments in the future. Now it is passed only to the features that are not yet implemented in the new `periodsModel` (that is filtering and exporting data - all of the `applicableDataActions` are meant to be handled within the `periodsModel` in the future).

File: app/public/views/periods/table/periodsTableHeader.js [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+ *
+ * In applying this license CERN does not waive the privileges and immunities
+ * granted to it by virtue of its status as an Intergovernmental Organization
+ * or submit itself to any jurisdiction.
+ */
+
+import { h } from '/js/src/index.js';
+import { getHeaderSpecial, headerSpecPresent, nonDisplayable } from '../../userView/data/headersSpecials.js';
+
+export default function periodsTableHeader(pageName, visibleFields, data, dataAccessModel) {
```
Comment (Ehevi) [inline app/public/views/periods/table/periodsTableHeader.js]: `dataAccessModel` is used for getting `headerSpecials` and is needed for retro-compatibility (handling navigation to other views that use the old model)

File: app/public/views/periods/overview/dataExport.js [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+        case exportPreferences.notSelected:
+            /* */
+            break;
+        case exportPreferences.visible:
+            /* */
+            break;
+        default:
+            break;
+    }
+};
```
Comment (xsalonx) [inline app/public/views/periods/overview/dataExport.js:91]: not dead code, but effectively do nothing

File: app/public/views/periods/overview/dataExport.js [hunk 8]
```diff
@@ -0,0 +1,157 @@
...
```
Comment (xsalonx) [inline app/public/views/periods/overview/dataExport.js:70]: dead code

File: test/public/components/itemsCounter.test.js [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+        }
+        const expectedResult = '41-50 of 57';
+
+        it('should not return null', () => {
+            assert.isNotNull(itemsCounter(mockPaginationModel));
+        });
+
+        it('should count the items as expected', () => {
+            assert.equal(itemsCounter(mockPaginationModel), expectedResult);
+        });
```
Comment (xsalonx) [inline test/public/components/itemsCounter.test.js:33]: considering second test, first one is redundant, in particular `itemsCounter` cannot return null at most can throw an error.

**Review by xsalonx — CHANGES_REQUESTED** [review]

File: app/public/utils/fetch/jsonFetch.js [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+ */
+
+import { fetchClient } from '/js/src/index.js';
+
+/**
+ * Send a request to a remote endpoint, and extract the response. If errors occurred, an array of errors containing title and detail is thrown
+ *
+ * The endpoint is expected to follow some conventions:
+ *   - If request is valid but no data must be sent, it must return a 204
+ *   - If data must be returned, the json response must contain a top level "data" key
```
[the remark this answers is not in the document]
Comment (Ehevi) [reply app/public/utils/fetch/jsonFetch.js]: Yes, thank you for spotting that.

File: app/public/components/table/pageSelector.js [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+        currentPage < pagesCount
+            ? pageIconButton(
+                currentPage + 1,
+                h('.forward-15-primary'),
+            )
+            : '',
+
+        // Move to the last site
+        currentPage < pagesCount
+            ? pageIconButton(
```
[the remark this answers is not in the document]
Comment (Ehevi) [reply app/public/components/table/pageSelector.js]: I agree, I've fixed that already.

File: app/public/views/periods/overview/dataExport.js (hunk 8 above)
Comment (Ehevi) [reply app/public/views/periods/overview/dataExport.js]: Right, that should come in the next PR when the handling functions are implemented

File: test/public/components/itemsCounter.test.js (hunk 9 above)
Comment (Ehevi) [reply test/public/components/itemsCounter.test.js]: removed

**Review by xsalonx — APPROVED** [review]

**Pull request merged**